File: pkg/routes/api/v1/pomodoro.go
package v1

import (
	"net/http"
	"strconv"
	"time"

	"github.com/labstack/echo/v4"
	"github.com/tecknohow/TaskDream/pkg/db"
	"github.com/tecknohow/TaskDream/pkg/models"
)

func ListPomodoroSessions(c echo.Context) error {
	userID := c.Get("user_id")
	if userID == nil {
		return c.JSON(http.StatusUnauthorized, map[string]string{"error": "unauthorized"})
	}

	taskIDStr := c.QueryParam("task_id")
	dateStr := c.QueryParam("date")

	var sessions []models.PomodoroSession
	query := db.Engine.Where("user_id = ?", userID)

	if taskIDStr != "" {
		taskID, _ := strconv.ParseInt(taskIDStr, 10, 64)
		query = query.And("task_id = ?", taskID)
	}

	if dateStr != "" {
		date, err := time.Parse("2006-01-02", dateStr)
		if err == nil {
			nextDay := date.Add(24 * time.Hour)
			query = query.And("started_at >= ? AND started_at < ?", date, nextDay)
		}
	}

	err := query.Desc("created").Find(&sessions)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": "database error"})
	}

	if sessions == nil {
		sessions = []models.PomodoroSession{}
	}

	return c.JSON(http.StatusOK, sessions)
}

func StartPomodoro(c echo.Context) error {
	userID := c.Get("user_id")
	if userID == nil {
		return c.JSON(http.StatusUnauthorized, map[string]string{"error": "unauthorized"})
	}

	var req struct {
		TaskID   int64 `json:"task_id"`
		Duration int   `json:"duration"` // in seconds, 0 means use settings default
	}

	if err := c.Bind(&req); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid request"})
	}

	// Get user's pomodoro settings
	settings := &models.PomodoroSettings{}
	has, _ := db.Engine.Where("user_id = ?", userID).Get(settings)
	if !has {
		settings.WorkDuration = 1500 // 25 min default
		settings.ShortBreak = 300
	}

	duration := req.Duration
	if duration == 0 {
		duration = settings.WorkDuration
	}

	now := time.Now()
	session := &models.PomodoroSession{
		TaskID:        req.TaskID,
		UserID:        userID.(int64),
		Duration:      duration,
		BreakDuration: settings.ShortBreak,
		Status:        "running",
		StartedAt:     &now,
	}

	_, err := db.Engine.Insert(session)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": "failed to start pomodoro"})
	}

	return c.JSON(http.StatusCreated, session)
}

func CompletePomodoro(c echo.Context) error {
	id, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid session id"})
	}

	now := time.Now()
	session := &models.PomodoroSession{
		Status:      "completed",
		CompletedAt: &now,
	}

	_, err = db.Engine.ID(id).Cols("status", "completed_at").Update(session)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": "failed to complete pomodoro"})
	}

	// Fetch the updated session
	updated := &models.PomodoroSession{}
	db.Engine.ID(id).Get(updated)

	return c.JSON(http.StatusOK, updated)
}

func CancelPomodoro(c echo.Context) error {
	id, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid session id"})
	}

	session := &models.PomodoroSession{
		Status: "cancelled",
	}

	_, err = db.Engine.ID(id).Cols("status").Update(session)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": "failed to cancel pomodoro"})
	}

	return c.JSON(http.StatusOK, map[string]string{"message": "pomodoro cancelled"})
}

func GetPomodoroSettings(c echo.Context) error {
	userID := c.Get("user_id")
	if userID == nil {
		return c.JSON(http.StatusUnauthorized, map[string]string{"error": "unauthorized"})
	}

	settings := &models.PomodoroSettings{}
	has, err := db.Engine.Where("user_id = ?", userID).Get(settings)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": "database error"})
	}

	if !has {
		// Return defaults
		settings = &models.PomodoroSettings{
			UserID:            userID.(int64),
			WorkDuration:      1500,
			ShortBreak:        300,
			LongBreak:         900,
			LongBreakInterval: 4,
		}
	}

	return c.JSON(http.StatusOK, settings)
}

func UpdatePomodoroSettings(c echo.Context) error {
	userID := c.Get("user_id")
	if userID == nil {
		return c.JSON(http.StatusUnauthorized, map[string]string{"error": "unauthorized"})
	}

	var req struct {
		WorkDuration      int  `json:"work_duration"`
		ShortBreak        int  `json:"short_break"`
		LongBreak         int  `json:"long_break"`
		LongBreakInterval int  `json:"long_break_interval"`
		AutoStartBreaks   bool `json:"auto_start_breaks"`
		AutoStartPomodoro bool `json:"auto_start_pomodoro"`
	}

	if err := c.Bind(&req); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid request"})
	}

	// Upsert
	existing := &models.PomodoroSettings{}
	has, _ := db.Engine.Where("user_id = ?", userID).Get(existing)

	settings := &models.PomodoroSettings{
		UserID:            userID.(int64),
		WorkDuration:      req.WorkDuration,
		ShortBreak:        req.ShortBreak,
		LongBreak:         req.LongBreak,
		LongBreakInterval: req.LongBreakInterval,
		AutoStartBreaks:   req.AutoStartBreaks,
		AutoStartPomodoro: req.AutoStartPomodoro,
	}

	if has {
		_, err := db.Engine.Where("user_id = ?", userID).Update(settings)
		if err != nil {
			return c.JSON(http.StatusInternalServerError, map[string]string{"error": "failed to update settings"})
		}
	} else {
		_, err := db.Engine.Insert(settings)
		if err != nil {
			return c.JSON(http.StatusInternalServerError, map[string]string{"error": "failed to create settings"})
		}
	}

	return c.JSON(http.StatusOK, settings)
}

func GetPomodoroStats(c echo.Context) error {
	userID := c.Get("user_id")
	if userID == nil {
		return c.JSON(http.StatusUnauthorized, map[string]string{"error": "unauthorized"})
	}

	dateStr := c.QueryParam("date")
	date := time.Now()
	if dateStr != "" {
		parsed, err := time.Parse("2006-01-02", dateStr)
		if err == nil {
			date = parsed
		}
	}

	startOfDay := time.Date(date.Year(), date.Month(), date.Day(), 0, 0, 0, 0, date.Location())
	endOfDay := startOfDay.Add(24 * time.Hour)

	// Count today's completed pomodoros
	todayCompleted, _ := db.Engine.Where("user_id = ? AND status = 'completed' AND started_at >= ? AND started_at < ?",
		userID, startOfDay, endOfDay).Count(&models.PomodoroSession{})

	// Count total time in completed pomodoros today
	var sessions []models.PomodoroSession
	db.Engine.Where("user_id = ? AND status = 'completed' AND started_at >= ? AND started_at < ?",
		userID, startOfDay, endOfDay).Find(&sessions)

	totalSeconds := 0
	for _, s := range sessions {
		totalSeconds += s.Duration
	}

	// Weekly count
	startOfWeek := startOfDay.AddDate(0, 0, -int(startOfDay.Weekday()))
	weeklyCompleted, _ := db.Engine.Where("user_id = ? AND status = 'completed' AND started_at >= ?",
		userID, startOfWeek).Count(&models.PomodoroSession{})

	return c.JSON(http.StatusOK, map[string]interface{}{
		"today_completed":  todayCompleted,
		"today_total_time": totalSeconds,
		"weekly_completed": weeklyCompleted,
	})
}

File: frontend/src/stores/pomodoro.ts
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import { pomodoroAPI } from '@/api/pomodoro'
import type { PomodoroSession, PomodoroSettings, PomodoroStats } from '@/types/models'

export const usePomodoroStore = defineStore('pomodoro', () => {
  const currentSession = ref<PomodoroSession | null>(null)
  const settings = ref<PomodoroSettings>({
    user_id: 0,
    work_duration: 1500,
    short_break: 300,
    long_break: 900,
    long_break_interval: 4,
    auto_start_breaks: false,
    auto_start_pomodoro: false
  })
  const stats = ref<PomodoroStats>({
    today_completed: 0,
    today_total_time: 0,
    weekly_completed: 0
  })
  const sessions = ref<PomodoroSession[]>([])
  const loading = ref(false)
  const error = ref<string | null>(null)

  // Timer state
  const elapsedSeconds = ref(0)
  const isOnBreak = ref(false)
  const breakSecondsLeft = ref(0)
  let timerInterval: ReturnType<typeof setInterval> | null = null
  let breakInterval: ReturnType<typeof setInterval> | null = null

  const isRunning = computed(() => currentSession.value?.status === 'running')

  const timeRemaining = computed(() => {
    if (!currentSession.value) return 0
    return Math.max(0, currentSession.value.duration - elapsedSeconds.value)
  })

  const formattedTimeRemaining = computed(() => {
    const total = isOnBreak.value ? breakSecondsLeft.value : timeRemaining.value
    const minutes = Math.floor(total / 60)
    const seconds = total % 60
    return `${String(minutes).padStart(2, '0')}:${String(seconds).padStart(2, '0')}`
  })

  const progress = computed(() => {
    if (!currentSession.value) return 0
    return (elapsedSeconds.value / currentSession.value.duration) * 100
  })

  const fetchSettings = async () => {
    try {
      const response = await pomodoroAPI.getSettings()
      settings.value = response.data
    } catch (err: any) {
      error.value = err.response?.data?.error || 'Failed to fetch settings'
    }
  }

  const updateSettings = async (newSettings: Partial<PomodoroSettings>) => {
    loading.value = true
    try {
      const response = await pomodoroAPI.updateSettings(newSettings)
      settings.value = response.data
    } catch (err: any) {
      error.value = err.response?.data?.error || 'Failed to update settings'
    } finally {
      loading.value = false
    }
  }

  const fetchStats = async () => {
    try {
      const response = await pomodoroAPI.getStats()
      stats.value = response.data
    } catch (err: any) {
      error.value = err.response?.data?.error || 'Failed to fetch stats'
    }
  }

  const fetchSessions = async (params?: { task_id?: number; date?: string }) => {
    loading.value = true
    try {
      const response = await pomodoroAPI.getSessions(params)
      sessions.value = response.data
    } catch (err: any) {
      error.value = err.response?.data?.error || 'Failed to fetch sessions'
    } finally {
      loading.value = false
    }
  }

  const startPomodoro = async (taskId: number) => {
    loading.value = true
    error.value = null

    try {
      const response = await pomodoroAPI.start(taskId, settings.value.work_duration)
      currentSession.value = response.data
      elapsedSeconds.value = 0
      isOnBreak.value = false

      if (timerInterval) clearInterval(timerInterval)
      timerInterval = setInterval(() => {
        elapsedSeconds.value++
        if (elapsedSeconds.value >= (currentSession.value?.duration || 0)) {
          completePomodoro()
        }
      }, 1000)

      return response.data
    } catch (err: any) {
      error.value = err.response?.data?.error || 'Failed to start pomodoro'
      throw error.value
    } finally {
      loading.value = false
    }
  }

  const completePomodoro = async () => {
    if (!currentSession.value) return

    if (timerInterval) {
      clearInterval(timerInterval)
      timerInterval = null
    }

    try {
      await pomodoroAPI.complete(currentSession.value.id)
      stats.value.today_completed++
      stats.value.today_total_time += currentSession.value.duration

      // Start break
      const breakDuration = (stats.value.today_completed % settings.value.long_break_interval === 0)
        ? settings.value.long_break
        : settings.value.short_break

      isOnBreak.value = true
      breakSecondsLeft.value = breakDuration

      if (breakInterval) clearInterval(breakInterval)
      breakInterval = setInterval(() => {
        breakSecondsLeft.value--
        if (breakSecondsLeft.value <= 0) {
          endBreak()
        }
      }, 1000)

      currentSession.value = null
      elapsedSeconds.value = 0
    } catch (err: any) {
      error.value = err.response?.data?.error || 'Failed to complete pomodoro'
    }
  }

  const cancelPomodoro = async () => {
    if (!currentSession.value) return

    if (timerInterval) {
      clearInterval(timerInterval)
      timerInterval = null
    }

    try {
      await pomodoroAPI.cancel(currentSession.value.id)
      currentSession.value = null
      elapsedSeconds.value = 0
    } catch (err: any) {
      error.value = err.response?.data?.error || 'Failed to cancel pomodoro'
    }
  }

  const endBreak = () => {
    if (breakInterval) {
      clearInterval(breakInterval)
      breakInterval = null
    }
    isOnBreak.value = false
    breakSecondsLeft.value = 0
  }

  const skipBreak = () => {
    endBreak()
  }

  return {
    currentSession,
    settings,
    stats,
    sessions,
    loading,
    error,
    elapsedSeconds,
    isOnBreak,
    breakSecondsLeft,
    isRunning,
    timeRemaining,
    formattedTimeRemaining,
    progress,
    fetchSettings,
    updateSettings,
    fetchStats,
    fetchSessions,
    startPomodoro,
    completePomodoro,
    cancelPomodoro,
    skipBreak
  }
})
